### PR TITLE
feat: adds urlpattern to ada

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
TL;DR

**WIP!**

----
So i started to implement https://wicg.github.io/urlpattern/
Seems that this is an work-in-progress spec, so maybe we'll have some freestyle stuff along the way of the implementation. 
I'm using https://github.com/kenchris/urlpattern-polyfill and https://github.com/denoland/rust-urlpattern as references. 

Don't know if there is a "right way" to do this, but I get the unicode identifier start/part ranges using the script:
```js
"use strict";

const fs = require("fs");

const regexIdentifierStart = /[$_\p{ID_Start}]/u;

function isValidIdStart(num) {
  const c = String.fromCharCode(num);
  return regexIdentifierStart.test(c);
}

const validRanges = [];
let start = null;

for (let c = 256; c < 0x10ffff; c++) {
  const isValid = isValidIdStart(c);
  if (isValid && start === null) {
    start = c;
  } else if (!isValid && start !== null) {
    validRanges.push([start, c - 1]);
    start = null;
  }
}

const writeRangeInFile = (filename, ranges) => {
  let file = "";
  for (const [base, upper] of ranges) {
    file += `\{${base}, ${upper}\},\n`;
  }
  fs.writeFile(filename, file, (err) => {});
};

console.log(validRanges.length);

writeRangeInFile("id_start_ranges.txt", validRanges);
```
++ something similar for the identifier-part with the regex ```/[$_\u200C\u200D\p{ID_Continue}]/u```
I got those regexes in https://github.com/kenchris/urlpattern-polyfill/blob/main/src/path-to-regex-modified.ts#LL70C1-L70C50

Then I used the ranges to create  bitwise masks in the unicode.cpp. 

Nothing is tested yet

It will took some time to finish! 

ref https://github.com/nodejs/node/issues/40844

### CODE REVIEW

File: include/ada/urlpattern.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+namespace ada {
+
+ada_really_inline bool is_valid_name_code_point(const char32_t& c,
+                                                bool is_first) noexcept;
+
+struct urlpattern_options {
+  bool ignore_case = false;
+};
+
+struct component_result {
```
Comment (anonrig) [inline include/ada/urlpattern.h:20]: `ada::component_result` seems a really vague naming, since it might be mistaken with `ada::url_components`

File: include/ada/urlpattern.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  component_result search;
+  component_result hash;
+  urlpattern_input input[];
+};
+
+struct urlpattern {
+  urlpattern(urlpattern_input input, std::string_view base_url,
+             std::optional<urlpattern_options> options);
+
+  urlpattern(std::optional<urlpattern_input> input,
```
Comment (anonrig) [inline include/ada/urlpattern.h:60]: what happens if you pass `std::nullopt, std::nullopt`, should we make both of these parameters optional?

File: src/urlpattern.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  const auto error_or_invalid = [&](const std::string_view msg, size_t start,
+                                    size_t end) {
+    if (policy != POLICY::LENIENT) {
+      throw std::invalid_argument(std::string(msg));
+    }
+    tokens.push_front({TOKEN_TYPE::INVALID_CHAR, start, end});
+  };
+
+  size_t index = 0;
+  while (index < input_size) {
```
Comment (anonrig) [inline src/urlpattern.cpp:52]: Can we iterate through a while loop and find the next available item through something like `buffer.find_first_of("*+?\\{}:")`?

File: include/ada/unicode.h [hunk 4]
```diff
@@ -201,6 +201,10 @@ ada_really_inline size_t percent_encode_index(const std::string_view input,
  * Return true if the content was ASCII.
  */
 constexpr bool to_lower_ascii(char* input, size_t length) noexcept;
+
+bool is_valid_identifier_part(const char32_t& c) noexcept;
```
Comment (anonrig) [inline include/ada/unicode.h:205]: I recommend adding comments to every function you create to help us review it better.

File: src/urlpattern.cpp (hunk 3 above)
Comment (miguelteixeiraa) [reply src/urlpattern.cpp:52]: Great idea!

File: include/ada/unicode.h (hunk 4 above)
Comment (miguelteixeiraa) [reply include/ada/unicode.h:205]: Sure thing!

File: include/ada/urlpattern.h (hunk 2 above)
Comment (miguelteixeiraa) [reply include/ada/urlpattern.h:60]: I'm not there yet, but this is one of the valid constructors according to the specs:

[constructor](https://wicg.github.io/urlpattern/#dom-urlpattern-urlpattern-input-options)(optional [URLPatternInput](https://wicg.github.io/urlpattern/#typedefdef-urlpatterninput) input = {}, optional [URLPatternOptions](https://wicg.github.io/urlpattern/#dictdef-urlpatternoptions) options = {});

I'll figure it out soon

File: include/ada/urlpattern.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+#include "ada/urlpattern_base.h"
+
+#include <unordered_map>
+#include <string_view>
+#include <optional>
+
+namespace ada::urlpattern {
+struct urlpattern_component_result {
+  std::string_view input;
+  std::unordered_map<std::string_view, std::optional<std::string_view>> groups;
```
Comment (anonrig) [inline include/ada/urlpattern.h:15]: @lemire How's the performance comparison of using std::optional in here versus std::variant<std::nullopt, std::string_view>?

File: include/ada/urlpattern_base.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+#define ADA_URLPATTERN_BASE_H
+
+#include "ada/helpers.h"
+
+#include <string_view>
+
+namespace ada::urlpattern {
+
+struct urlpattern_options {
+  std::string_view delimiter = "";
```
Comment (anonrig) [inline include/ada/urlpattern_base.h:11]:
```suggestion
  std::string_view delimiter{};
```

File: include/ada/urlpattern_constructor_string_parser.h [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+  // https://wicg.github.io/urlpattern/#make-a-component-string
+  ada_really_inline std::u32string_view make_component_string();
+
+  /// https://wicg.github.io/urlpattern/#next-is-authority-slashes
+  ada_really_inline bool next_is_authority_slashes();
+
+  std::u32string_view input;
+  std::vector<token> token_list;
+  size_t component_start = 0;
```
Comment (anonrig) [inline include/ada/urlpattern_constructor_string_parser.h:90]:
```suggestion
  size_t component_start{0};
```

File: include/ada/urlpattern_internals.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+#define ADA_URLPATTERN_INTERNALS_H
+
+#include "ada/urlpattern_base.h"
+#include <string_view>
+#include "regex"
+#include "vector"
+
+namespace ada::urlpattern {
+// https://wicg.github.io/urlpattern/#component
+struct urlpattern_component {
```
Comment (anonrig) [inline include/ada/urlpattern_internals.h:11]: Since this is already under ada::urlpattern namespace, why not call this struct component?

File: src/urlpattern_canonicalization.cpp [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  // followed by "://dummy.test", with dummyURL as url.
+
+  // TODO: make it cheaper
+  std::u32string url = std::u32string(protocol) + U"://dummy.test";
+
+  auto utf8_size = ada::idna::utf8_length_from_utf32(url.data(), url.size());
+  std::string final_utf8_url(utf8_size, '\0');
+  ada::idna::utf32_to_utf8(url.data(), url.size(), final_utf8_url.data());
+
+  if (ada::can_parse(final_utf8_url)) {
```
Comment (anonrig) [inline src/urlpattern_canonicalization.cpp:26]: Can you add a todo in here to optimize the can_parse function?

File: src/urlpattern_constructor_string_parser.cpp [hunk 10]
```diff
@@ -0,0 +1,579 @@
...
+        }
+        break;
+      }
+      case PARSER_STATE::HASH: {
+        // do nothing;
+        break;
+      }
+      case PARSER_STATE::DONE: {
+        // 1. Assert: This step is never reached.
+        assert(false);
```
Comment (anonrig) [inline src/urlpattern_constructor_string_parser.cpp:292]: We have ada development asserts that is only enabled on development builds. @lemire write a good blog post about this: https://lemire.me/blog/2023/03/15/runtime-asserts-are-not-free/?amp

File: src/urlpattern_constructor_string_parser.cpp [hunk 11]
```diff
@@ -0,0 +1,579 @@
...
+    return false;
+  }
+
+  // 2. If the result of running is a non-special pattern char given parser,
+  // parser’s token index + 2, and "/" is false, then return false.
+  if (!is_nonspecial_pattern_char(token_index + 2, U"/")) {
+    return false;
+  }
+
+  return true;
```
Comment (anonrig) [inline src/urlpattern_constructor_string_parser.cpp:462]: This whole function can be a single boolean expression.

File: src/urlpattern_constructor_string_parser.cpp [hunk 12]
```diff
@@ -0,0 +1,579 @@
...
+  //   token’s type is "char";
+  //   token’s type is "escaped-char"; or
+  //   token’s type is "invalid-char",
+  //
+  // then return true.
+
+  if (safe_token->type == TOKEN_TYPE::CHAR ||
+      safe_token->type == TOKEN_TYPE::ESCAPED_CHAR ||
+      safe_token->type == TOKEN_TYPE::INVALID_CHAR) {
+    return true;
```
Comment (anonrig) [inline src/urlpattern_constructor_string_parser.cpp:488]: This if statement can be reduced into the return value.

File: src/urlpattern_constructor_string_parser.cpp [hunk 13]
```diff
@@ -0,0 +1,579 @@
...
+
+// https://wicg.github.io/urlpattern/#get-a-safe-token
+ada_really_inline token *constructor_string_parser::get_safe_token(
+    size_t &index) {
+  // 1. If index is less than parser’s token list's size, then return parser’s
+  // token list[index].
+  if (index < token_list.size()) return &token_list[index];
+  // 2. Assert: parser’s token list's size is greater than or equal to 1.
+  // TODO: messages for the asserts? conditional and then throw error?
+  assert(token_list.size() >= 1);
```
Comment (anonrig) [inline src/urlpattern_constructor_string_parser.cpp:502]: one wise man said "asserts are not free"

File: src/urlpattern_constructor_string_parser.cpp [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+  // 5. Assert: token’s type is "end".
+  assert(token_list[last_index].type == TOKEN_TYPE::END);
+  return &token_list[last_index];
+}
+
+// https://wicg.github.io/urlpattern/#is-a-group-open
+ada_really_inline bool constructor_string_parser::is_group_open() {
+  // If parser’s token list[parser’s token index]'s type is "open", then
+  // return true. Else return false.
+  return token_list[token_index].type == TOKEN_TYPE::OPEN;
```
Comment (anonrig) [inline src/urlpattern_constructor_string_parser.cpp:515]: This is not safe. We should add development asserts to make sure token_index is smaller than token_list length.

File: src/urlpattern_pattern_parser.cpp [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+    std::u32string_view input) {
+  // 1. Assert: input is an ASCII string.
+  assert(ada::idna::is_ascii(input));
+
+  std::u32string result = U"";
+  size_t index = 0;
+  while (index < input.size()) {
+    size_t pos = input.find_first_of(U".+*?^${}()[]|/\\)");
+    if (pos == std::string_view::npos) {
+      result = result += input.substr(index, input.size());
```
Comment (anonrig) [inline src/urlpattern_pattern_parser.cpp:30]: This line is weird.

File: src/urlpattern_pattern_parser.cpp [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+    type = PART_TYPE::SEGMENT_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
+  }
+  // 12. Else if regexp value is the full wildcard regexp value:
+  else if (regexp_value == full_wildcard_regexp_value) {
+    // 1. Set type to "full-wildcard".
+    type = PART_TYPE::FULL_WILDCARD;
+    // 2. Set regexp value to the empty string.
+    regexp_value.clear();
```
Comment (anonrig) [inline src/urlpattern_pattern_parser.cpp:204]: We are assigning and later clearing this value, which is creating performance degregation. We should eventually optimize it to reduce unnecessary allocations.

File: src/urlpattern_pattern_parser.cpp [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+  // and "other-modifier".
+  std::optional<token *> t = try_to_consume_token(TOKEN_TYPE::OTHER_MODIFIER);
+
+  // 2. If token is not null, then return token.
+  if (t.has_value()) {
+    return t;
+  }
+
+  // 3. Set token to the result of running try to consume a token given parser
+  // and "asterisk".
```
Comment (anonrig) [inline src/urlpattern_pattern_parser.cpp:247]: there was a function like t.value_or("default val")

File: src/urlpattern_pattern_parser.cpp [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+pattern_parser::try_to_consume_regexp_or_wildcard_token(
+    std::optional<token *> &name_token) {
+  // Let token be the result of running try to consume a token given parser and
+  // "regexp".
+  std::optional<token *> regexp_or_wildcard =
+      try_to_consume_token(TOKEN_TYPE::REGEXP);
+
+  // If name token is null and token is null, then set token to the result of
+  // running try to consume a token given parser and "asterisk".
+  if (!name_token.has_value() && !regexp_or_wildcard.has_value()) {
```
Comment (anonrig) [inline src/urlpattern_pattern_parser.cpp:262]: There is fast path here. If name_token does not have a value, you dont need to try to consume token right?

**Comment by bricss** [comment]
Is there any way to crank 🔧 this up for https://github.com/nodejs/node/issues/51060 needs? 🤔

**Comment by bricss** [comment]
Houston, do you read me? 📻

File: include/ada/urlpattern.h (hunk 5 above)
Comment (lemire) [reply include/ada/urlpattern.h:15]: I don't expect it matters much.

**Comment by lemire** [comment]
@bricss Are you available to help push this forward?

**Comment by bricss** [comment]
Yes, I'm like 24/7 into it, with only one exception, I'm not great C++ coding 🤷‍♂️ atm, but if someone willing give me a slight kick-in, I will code as mad rabbit 🐰 I promise 😬 (coz what I do in JS) 👾

**Comment by lemire** [comment]
@bricss Lack of knowledge of C++ could be a problem.
